packages/core/src/state-management/named-p-queue.ts

```diff
@@ -0,0 +1,63 @@
+import PQueue from 'p-queue';
+import { Mutex } from 'await-semaphore';
+
+/**
+ * Set of named PQueues.
+ * When a task is done, it checks for pending tasks. No pending tasks means PQueue for the name is cleared.
+ */
+export class NamedPQueue {
+  #lanes: Map<string, PQueue>;
+  #operations: Mutex = new Mutex();
+
+  constructor(lanes: Map<string, PQueue> = new Map()) {
+    this.#lanes = lanes;
+  }
+
+  /**
+   * Get a queue: return existing one, or create new.
+   *
+   * There can only be one `queue` or `remove` operation running at a time.
+   */
+  private queue(name: string): Promise<PQueue> {
+    return this.#operations.use(async () => {
+      const found = this.#lanes.get(name);
+      if (found) {
+        return found;
+      } else {
+        const queue = new PQueue({ concurrency: 1 });
+        this.#lanes.set(name, queue);
+        return queue;
+      }
+    });
+  }
+
+  /**
+   * Remove a queue. Called after a task is done.
+   *
+   * There can only be one `queue` or `remove` operation running at a time.
+   */
+  private remove(name: string): Promise<void> {
+    return this.#operations.use(async () => {
+      const found = this.#lanes.get(name);
+      if (found && found.size === 0) {
```

Comment:
this `.size` check looks racy to me.  Is there anything that prevents an item from being added or removed from the queue after this check has occurred?

Reply:
I think this would work if the `#operations` mutex were held not just when create/fetching the queue but also while the task is being added to it.  That way if you see the size is 0 here you know nothing can get added until the mutex is released.

Reply:
This is puzzling. JS execution model is a single-threaded event-loop with events [processed until completion](https://developer.mozilla.org/en-US/docs/Web/JavaScript/EventLoop#run-to-completion). Nothing can stop the execution here after the check, and start executing something else.

I wrapped addition, but what the case you are describing here is a total impossibility.

Reply:
Yep, you are right of course.  Still getting used to the Node concurrency model.